### PR TITLE
Feature/special price

### PR DESCRIPTION
**Pull request by ghost**
**Tasks:**
- [x] done
- [ ] todo
- [ ] got feedback

__Informations: special price calculations

| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| Tests pass? | yes |
| Fixed tickets | https://github.com/massiveart/POOL-ALPIN/issues/1149 |
| Doc | none |

### CODE REVIEW

File: Pricing/ItemPriceCalculator.php [hunk 1]
```diff
@@ -50,23 +50,28 @@ public function calculate($item, $currency = null, $useProductsPrice = true)
...
             $product = $item->getCalcProduct();
-            $price = $this->priceManager->getBulkPriceForCurrency($product, $item->getCalcQuantity(), $currency);
-
-            // no price set - return 0
-            if ($price === null) {
-                return 0;
+            $specialPrice = $this->priceManager->getSpecialPriceForCurrency($product, $currency);
+            if (!empty($specialPrice)) {
+                $priceValue = $specialPrice->getPrice();
+            } else {
```
Comment (turbo-ele) [inline Pricing/ItemPriceCalculator.php]: Please also consider that a bulk price can be smaller than the special price. So you need to check the special price and the bulk price and take the smaller of both.

File: Pricing/ItemPriceCalculator.php [hunk 2]
```diff
@@ -137,16 +142,27 @@ protected function validateNotNull($key, $value)
...
         $currency = $this->getCurrency($currency);
-        
+
+        // get bulk price
         if ($useProductPrice) {
             $product = $item->getCalcProduct();
-            $price = $this->priceManager->getBulkPriceForCurrency($product, $item->getCalcQuantity(), $currency);
-            $price = $price->getPrice();
+            $specialPrice = $this->priceManager->getSpecialPriceForCurrency($product, $currency);
+            if (!empty($specialPrice)) {
```
Comment (turbo-ele) [inline Pricing/ItemPriceCalculator.php]: see comment above

File: Pricing/ItemPriceCalculator.php [hunk 3]
```diff
@@ -50,23 +50,50 @@ public function calculate($item, $currency = null, $useProductsPrice = true)
         // get bulk price
         if ($useProductsPrice) {
             $product = $item->getCalcProduct();
-            $price = $this->priceManager->getBulkPriceForCurrency($product, $item->getCalcQuantity(), $currency);
+
+            //get special price
+            $specialPrice = $this->priceManager->getSpecialPriceForCurrency($product, $currency);
+            if (is_object($specialPrice)) {
```
Comment (turbo-ele) [inline Pricing/ItemPriceCalculator.php]: i  guess checking for an object is not really necessary
and try avoiding the else.

```
$specialPriceValue = null;
if ($specialPrice) {
    $specialPriceValue = $specialPrice->getPrice();
}
```

File: Pricing/ItemPriceCalculator.php [hunk 4]
```diff
@@ -50,23 +50,50 @@ public function calculate($item, $currency = null, $useProductsPrice = true)
...
+                $specialPriceValue = $specialPrice->getPrice();
+            } else {
+                $specialPriceValue = null;
+            }
+
+            //get bulk price
+            $bulkPrice = $this->priceManager->getBulkPriceForCurrency($product, $item->getCalcQuantity(), $currency);
+            if (is_object($bulkPrice)) {
```
Comment (turbo-ele) [inline Pricing/ItemPriceCalculator.php]: see comment above

File: Pricing/ItemPriceCalculator.php [hunk 5]
```diff
@@ -50,23 +50,50 @@ public function calculate($item, $currency = null, $useProductsPrice = true)
...
+                $bulkPriceValue = $bulkPrice->getPrice();
+            } else {
+                $bulkPriceValue = null;
+            }
+
+            //take the smallest
+            if(!empty($specialPriceValue) && !empty($bulkPriceValue)) {
```
Comment (turbo-ele) [inline Pricing/ItemPriceCalculator.php]: way to many conditions here! reduce!

File: Pricing/ItemPriceCalculator.php [hunk 6]
```diff
@@ -137,16 +164,48 @@ protected function validateNotNull($key, $value)
...
+
         if ($useProductPrice) {
             $product = $item->getCalcProduct();
-            $price = $this->priceManager->getBulkPriceForCurrency($product, $item->getCalcQuantity(), $currency);
-            $price = $price->getPrice();
-        } else {
-            $price = $item->getPrice();
+
+            //get special price
+            $specialPrice = $this->priceManager->getSpecialPriceForCurrency($product, $currency);
```
Comment (turbo-ele) [inline Pricing/ItemPriceCalculator.php]: duplicated code! 
put the logic above into a function and call the function from here

File: Pricing/ItemPriceCalculator.php (hunk 5 above)
Comment (turbo-ele) [inline Pricing/ItemPriceCalculator.php]:
```
$priceValue = $bulkPriceValue;
if ($specialPriceValue && $bulkPriceValue) {
    if ($specialPriceValue < $bulkPriceValue) {
        $priceValue = $specialPriceValue;
    }
} elseif ($specialPriceValue) {
    $priceValue = $specialPriceValue;
}

if ($priceValue) {
    return 0;
}
```

File: Pricing/ItemPriceCalculator.php [hunk 7]
```diff
@@ -31,6 +31,52 @@ public function __construct(
...
+            $bulkPriceValue = $bulkPrice->getPrice();
+        }
+
+        //take the smallest
+        if(!empty($specialPriceValue) && !empty($bulkPriceValue)) {
+            $priceValue = $specialPriceValue;
+            if ($specialPriceValue > $bulkPriceValue) {
+                $priceValue = $bulkPriceValue;
+            }
+        } else if (!empty($specialPriceValue)) {
```
Comment (turbo-ele) [inline Pricing/ItemPriceCalculator.php]: elseif

File: Pricing/ItemPriceCalculator.php [hunk 8]
```diff
@@ -49,24 +95,17 @@ public function calculate($item, $currency = null, $useProductsPrice = true)
 
         // get bulk price
         if ($useProductsPrice) {
-            $product = $item->getCalcProduct();
-            $price = $this->priceManager->getBulkPriceForCurrency($product, $item->getCalcQuantity(), $currency);
-
-            // no price set - return 0
-            if ($price === null) {
-                return 0;
```
Comment (turbo-ele) [inline Pricing/ItemPriceCalculator.php:57]: actually this is missing by calling the function. does it work correctly anyway?

Comment (ghost) [inline Pricing/ItemPriceCalculator.php:57]: it's replaced by a function, yes. It works, just tested

Comment (turbo-ele) [inline Pricing/ItemPriceCalculator.php:57]: yes, but it's not the same. the rest of the function would be skipped at this point with the return. 
but if it still works its fine

**Comment by turbo-ele** [comment]
ack

**Pull request merged**